### PR TITLE
[LiveComponent] Lazy load LiveComponent

### PR DESCRIPTION
**Pull request by smnandre**
# UPDATED COMMENT

This PR add loading support for LiveComponent

* `loading="defer"` to load the component on page load 
* `loading="lazy"`  to load the component when the component is visible 
 
 
```twig
{# HTML Syntax #}
<twig:HeavyComponent loading="lazy" /> 

{# Twig Syntax #}
{{ component('HeavyComponent, { loading: "lazy" }) }}
```

The `defer`attribute is deprecated for this solution.

See more details in the document.

--- 

# ORIGINAL PR:

I have some live component stuff in progress, so let's post this one now to start the discussion.

This PR implements a _volontarely-minimal_ `loading="lazy"` mode on LiveComponents, based on top of the `defer` one.

## ⚠️  Live Component 

This feature is dedicated to Live Component (as there is some particularities due to keeping "state").

(if you need some lazy loading with classic twig components, there are already methods out there to handle such things) 

## Defer ? Lazy ? 

Both defer and lazy attributes allow to ... defer the full rendering of the component. So when the page first renders, it contains only a skeletton / empty component.

Then, depending of the attribute used, the component will fetch its content via a fetch request:

* `defer`:  when the page loads (it react to the "connect" Stimulus)
* `lazy`: when the component enters the viewport (so as soon it's at least 1px **visible** in it)

I precise "**visible**" because if you use some tabbed content,  horizontal sliders, etc... the "intersect" is not triggered until the component is actually.... visible. 

## Usage

```twig
<twig:Acme foo="bar" lazy />
```

Note: `lazy` implies `defer`, so this is equivalent to 

```twig
<twig:Acme foo="bar" defer lazy />
```

## So.. defer or lazy ? 

As i see it:
* `defer` should be used for any content that will 100% be visible in the main content, but too heavy to be computed with the original request (think Facebook/Insta first results of the timeline)
* `Lazy` should be used for anything really not in the viewport, and that you're not sure will be rendered to the user (comments, tabbed content, ...)

In both situation, this is not a good idea to have too many lazy live components in the page, it's probably a sign that you should use Turbo instead!

(i'll add some doc about defer/lazy and layout shift too, but the idea is: better "keep the space you'll need")

### IntersectionObserver

The loading is triggered with a dedicated intersection observer.

It happens only once and the the element in unregistered.

No settings allowed (as images, iframes, etc.. in the browser)

TODO: optimize threesholds.

### Loading=lazy

I chose to use the `loading="lazy"`  attribute tag on the rendered HTML, as the goal is to have the same behaviour on the browser side.

```html
<div
    loading="lazy"
    data-....
    class="foo"
>
```

It is standard for image, iframes, maybe portals soon, and i think it's self-explanatory.

Example: https://web.dev/articles/browser-level-image-lazy-loading

For browsers and for seo , this seems to be a good way to hint what this div is (i'll see aria stuff later as it touches many things on the codebase)

### Boolean attribute values

This is how we document the defer feature

```twig
{# expression can be a bool, a var, .... #}
{{ component('acme', {foo: "bar, defer: expression}) }}
```

For better DX, i add a small bool cast for both lazy and defer attributes values. It allows to have some dynamism on the calling side with HTML syntax too.

```twig
{# Defer: true #}
<twig:Acme foo="bar" defer />
<twig:Acme foo="bar" defer="defer" />
<twig:Acme foo="bar" defer="true" />
<twig:Acme foo="bar" defer="{{ 8 < 42 }}" />
<twig:Acme foo="bar" :defer="true" />

{# Also defer: true ! #} 
<twig:Acme foo="bar" defer="false" />

{# Defer: false #}
<twig:Acme foo="bar" defer="" />
<twig:Acme foo="bar" defer="{{ 8 > 42 }}" />
<twig:Acme foo="bar" :defer="false" />
```

(Tests / doc incoming, demo(s) in progress)

-- IN PROGRESS --

### CODE REVIEW

**Comment by 94noni** [comment]
I use on some projects https://www.stimulus-components.com/docs/stimulus-content-loader/
and I would love to have this in symfony/ux 👍🏻

**Comment by kbond** [comment]
To clarify?
1. defer = on page load
2. lazy = when in view

**Comment by WebMamba** [comment]
If I follow the logic:
for every component with a lazy attribute, we add the element into the IntersectionObserver.
So now when a component is on the viewport the IntersionObserver executes a callback that finds all the components present in the viewport and dispatches an event 'live:appear' for each of them. 

Based on what I understand I have a question:
Can I trigger this live:appear before it's trigger by IntersionObserver without having the IntersionObjerver to trigger a live:appear a second time? 
Do you see what I mean?

**Comment by smnandre** [comment]
> lazy = when in view

Yep i'll edit the description you're right :)

**Comment by smnandre** [comment]
> So now when a component is on the viewport the IntersionObserver 

It's not "when a component is on the viewport" it can be a lot of things (a component is "no more intersecting" for instance)

So let's say "at a given time" for now :) 

> executes a callback that finds all the components present in the viewport and dispatches an event 'live:appear' for each of them.

Yes

> Can I trigger this live:appear before it's trigger by IntersionObserver

You can dispatch any event you want :) 

> without having the IntersionObjerver to trigger a live:appear a second time?

In that particular scenario, as the intersection observer is still obeserving the component, it would indeed trigger a "live:appear" event when the component start intersecting the viewport, and then unobserve it. 

Now my question is: why would you do that ? :) 

I mean, if you trigger manually some "disconnect" event to a stimulus component, while not really disconnecting it, that would create some weird side effects... same for all the "`live:*`" events in fact :) 

If someone voluntarily call internal/prive code, there is not much we can do to anticipate it / handle it, and there is nothing we can do to prevent it. 

Did you have a specific / particular case in mind ?

File: src/LiveComponent/doc/index.rst [hunk 1]
```diff
@@ -2226,6 +2226,9 @@ To validate only on "change", use the ``on(change)`` modifier:
 Deferring / Lazy Loading Components
 -----------------------------------
 
+Defer Rendering
```
Comment (weaverryan) [inline src/LiveComponent/doc/index.rst]: Before this header, let's add 1-2 sentences to quickly introduce defer vs lazy. I fear that the `defer` and `lazy` will be confusing, so let's very clearly give them the choice. Then they can read how to implement the one they want.

**Review by weaverryan — CHANGES_REQUESTED** [review]
That's quite an elegant, simple solution. Turbo also uses `loading="lazy"` on Turbo Frames, so I like that attribute. 

Though, then doing `<twig:CustomThing loading="lazy">` would be more consistent than having a `lazy` boolean attribute. Doing `loading="lazy"` might also help clarify `loading` vs `defer`. The explanation would be something like:

> To defer loading your component until later, add the `defer` attribute. The component will render empty (or with some loading markup) then immediately make an AJAX call to load the real content.
> To not load the content until the element becomes visible, add `loading="lazy"`.

So, like Turbo, there would be `loading="eager"` (the default) and `loading="lazy"`. And we expose this detail to the user vs hiding it from them.

**Comment by kbond** [comment]
> Doing loading="lazy" might also help clarify loading vs defer. The explanation would be something like:

Should we then support `loading="defer"` to be consistent (and maybe deprecate the `defer` attribute)?

**Comment by smnandre** [comment]
> Should we then support `loading="defer"` to be consistent (and maybe deprecate the `defer` attribute)?

In HTML, `loading="defer"` does not exist. 
* `defer` attribute on scripts (in "opposition" to `async` )
*` loading="lazy"` on image, iframe ..

As lazy implies defer here, your suggestion makes perfect sense... but it complexifies a bit the DX

```twig
<twig:MyComponent defer lazy="{{ loop.index > 4 }}" />
```

```twig
<twig:MyComponent loading="{{ loop.index > 4 ? 'lazy'  : 'defer' }}" />
```

Can we keep "defer" and "lazy" as some kind of aliases of "loading=XXX" ? Or is it not worth it ?

File: src/LiveComponent/doc/index.rst (hunk 1 above)
Comment (smnandre) [reply src/LiveComponent/doc/index.rst]: I'll do it after we decide the "behaviour" / "naming", that will have an impact on the way to explain things. 

But it's noted !

**Comment by weaverryan** [comment]
Yea... *just* having a `loading` (set to `defer` or `lazy`), indeed simplifies things.

> Can we keep "defer" and "lazy" as some kind of aliases of "loading=XXX" ? Or is it not worth it ?

The `loading="{{ loop.index > 4 ? 'lazy'  : 'defer' }}"` is indeed longer, but the expression (`loop.index > 4`) is the same, so it feels like both have equal complexity. This longer version also still reads very clear. I'd say we do NOT keep these aliases.

**Comment by smnandre** [comment]
Updated for loading="lazy" / loading="defer" 
+ deprecate "defer" attribute
+ doc

**Review by kbond — APPROVED** [review]
Docs/backend code looks good!

**Comment by WebMamba** [comment]
Hey, I just have a question in a previous project we were using lazy loading in a specific way. The user was able to scroll a list of full-page images and we started the lazy load 3 images before the current image. I don't know if this PR can cover such a need. And if we should care about such a need?

**Comment by smnandre** [comment]
> Hey, I just have a question in a previous project we were using lazy loading in a specific way. The user was able to scroll a list of full-page images and we started the lazy load 3 images before the current image. I don't know if this PR can cover such a need. And if we should care about such a need?

I dont think this will be necessary. With a placeholder, and knowing your component will load probably in less than 200ms so it will more look like an animation than an ajax request. 

An image is often much larger than a live component request. 

I’d suggest we release without any customization and lets see if the need presents itself IRL ?

File: src/LiveComponent/doc/index.rst [hunk 2]
```diff
@@ -2236,37 +2236,92 @@ To validate only on "change", use the ``on(change)`` modifier:
...
+will be able to interact with it sooner, improving both the user experience
+and the overall performance of your application.
+
+.. note::
+
+    Behind the scenes, your component *is* created & mounted during the initial
+    page load, but it isn't rendered. So keep your heavy work to methods in
+    your component (e.g. ``getProducts()``) that are only called when rendering.
+
+Loading "defer"
```
Comment (weaverryan) [inline src/LiveComponent/doc/index.rst]:
```suggestion
Loading "defer" (Ajax on Load)
```

File: src/LiveComponent/doc/index.rst [hunk 3]
```diff
@@ -2236,37 +2236,92 @@ To validate only on "change", use the ``on(change)`` modifier:
...
+.. code-block:: twig
+
+    {# With the HTML syntax #}
+    {{ component('SomeHeavyComponent', { loading: 'defer' }) }}
 
 This renders an empty ``<div>`` tag, but triggers an Ajax call to render the
 real component once the page has loaded.
 
-.. note::
+Loading "lazy"
```
Comment (weaverryan) [inline src/LiveComponent/doc/index.rst]:
```suggestion
Loading "lazy" (Ajax when Visible)
```

File: src/LiveComponent/doc/index.rst [hunk 4]
```diff
@@ -2236,37 +2236,92 @@ To validate only on "change", use the ``on(change)`` modifier:
 Deferring / Lazy Loading Components
 -----------------------------------
 
+When a page loads, all components are rendered immediately. If a component is
+heavy to render, you can defer its rendering after the page has loaded.
```
Comment (weaverryan) [inline src/LiveComponent/doc/index.rst]:
```suggestion
heavy to render, you can defer its rendering until after the page has loaded. This is done
by making an Ajax call to load the component's real content either as soon as the page loads
(``defer``) or when the component becomes visible (``lazy``).
```

Then I think we can remove the next 2 paragraphs. Shorter text is more likely to be read.

File: src/LiveComponent/doc/index.rst [hunk 5]
```diff
@@ -2236,37 +2236,92 @@ To validate only on "change", use the ``on(change)`` modifier:
...
+
+By loading a component asynchronously, the page will load faster and the user
+will be able to interact with it sooner, improving both the user experience
+and the overall performance of your application.
+
+.. note::
+
+    Behind the scenes, your component *is* created & mounted during the initial
+    page load, but it isn't rendered. So keep your heavy work to methods in
+    your component (e.g. ``getProducts()``) that are only called when rendering.
```
Comment (weaverryan) [inline src/LiveComponent/doc/index.rst]: ... called from the component's template.

File: src/LiveComponent/doc/index.rst [hunk 6]
```diff
@@ -2236,37 +2236,92 @@ To validate only on "change", use the ``on(change)`` modifier:
...
+In this case, you can load them "lazily" when they are in the viewport.
+
+By loading a component asynchronously, the page will load faster and the user
+will be able to interact with it sooner, improving both the user experience
+and the overall performance of your application.
+
+.. note::
+
+    Behind the scenes, your component *is* created & mounted during the initial
+    page load, but it isn't rendered. So keep your heavy work to methods in
```
Comment (weaverryan) [inline src/LiveComponent/doc/index.rst]:
```suggestion
    page load, but its template isn't rendered. So keep your heavy work to methods in
```

File: src/LiveComponent/doc/index.rst [hunk 7]
```diff
@@ -2236,37 +2236,92 @@ To validate only on "change", use the ``on(change)`` modifier:
...
+
+.. code-block:: twig
+
+    {# With the Twig syntax #}
+    {{ component('SomeHeavyComponent', { loading: 'lazy' }) }}
+
+This renders an empty ``<div>`` tag. The real component is only rendered when
+it appears in the viewport.
+
+Defer or Lazy ?
```
Comment (weaverryan) [inline src/LiveComponent/doc/index.rst]:
```suggestion
Defer or Lazy?
```

I know, that's a style thing that varies by language, but without the space is the way in the Symfony docs.

File: src/LiveComponent/src/EventListener/DeferLiveComponentSubscriber.php [hunk 8]
```diff
@@ -31,11 +31,22 @@ final class DeferLiveComponentSubscriber implements EventSubscriberInterface
     public function onPostMount(PostMountEvent $event): void
     {
         $data = $event->getData();
+
         if (\array_key_exists('defer', $data)) {
-            $event->addExtraMetadata('defer', true);
+            trigger_deprecation('symfony/ux-live-component', '2.17', 'The "defer" attribute is deprecated and will be removed in 3.0. Use the "loading" attribute instead.');
```
Comment (weaverryan) [inline src/LiveComponent/src/EventListener/DeferLiveComponentSubscriber.php]:
```suggestion
            trigger_deprecation('symfony/ux-live-component', '2.17', 'The "defer" attribute is deprecated and will be removed in 3.0. Use the "loading" attribute instead set to the value "defer"');
```

File: src/LiveComponent/src/EventListener/DeferLiveComponentSubscriber.php [hunk 9]
```diff
@@ -31,11 +31,22 @@ final class DeferLiveComponentSubscriber implements EventSubscriberInterface
...
+            if ($data['defer']) {
+                $event->addExtraMetadata('loading', 'defer');
+            }
             unset($data['defer']);
         }
 
+        if (\array_key_exists('loading', $data)) {
+            if (\in_array($data['loading'], ['defer', 'lazy'], true)) {
```
Comment (weaverryan) [inline src/LiveComponent/src/EventListener/DeferLiveComponentSubscriber.php]: Ok, what if the user does `loading="foo"`? On the one hand, I think we should throw an exception. On the other hand, maybe we should allow that and render it as an attribute. Right now, I think we're doing neither - as it looks like `loading="foo"` would not be rendered due to the `unset()` below, right?

**Review by weaverryan — CHANGES_REQUESTED** [review]
This looks fantastic - a few last tweaks only!

File: src/LiveComponent/doc/index.rst (hunk 7 above)
Comment (smnandre) [reply src/LiveComponent/doc/index.rst]: 100%, just a french habit :)

File: src/LiveComponent/src/EventListener/DeferLiveComponentSubscriber.php (hunk 9 above)
Comment (smnandre) [reply src/LiveComponent/src/EventListener/DeferLiveComponentSubscriber.php]: The only HTML element that supports the `loading` attribute today are `img` and `iframe`, probably `portal` soon. 

None of them can be a LiveComponent so i think we can ignore this for now. And decide what to do if someone raises some concern about that ?

wdyt ?

File: src/LiveComponent/doc/index.rst (hunk 4 above)
Comment (smnandre) [reply src/LiveComponent/doc/index.rst]: Thank you, much better

**Comment by smnandre** [comment]
* Updated the doc following feedback
* Updated CHANGELOG.md
* Updated UPGRADE.md

All is ready for me :)

File: src/LiveComponent/src/EventListener/DeferLiveComponentSubscriber.php (hunk 9 above)
Comment (weaverryan) [reply src/LiveComponent/src/EventListener/DeferLiveComponentSubscriber.php]: In that case, throwing an exception would be the most helpful for the user and the most conservative (if we removed the exception and allowed the attribute to render later, there is no way that will affect anyone).

**Comment by weaverryan** [comment]
One last comment (hopefully) and a phpcs failure :)

**Comment by smnandre** [comment]
phpcs due to rebase, they are fixed in: https://github.com/symfony/ux/pull/1579

File: src/LiveComponent/src/EventListener/DeferLiveComponentSubscriber.php (hunk 9 above)
Comment (smnandre) [reply src/LiveComponent/src/EventListener/DeferLiveComponentSubscriber.php]: So we should ignore `false`  and `''` value, keep `lazy` or `defer`, and throw elsewhen.. right ?

**Comment by weaverryan** [comment]
Fantastic! Thanks Simon!

**Comment by smnandre** [comment]
Thanks to @jakubtobiasz for paving the way with https://github.com/symfony/ux/pull/1143 :)

**Pull request merged**